### ISSUE TITLE
tensorflow addon sampler example

### ISSUE THREAD

**Issue by ruiguo-bio** [ageron/handson-ml2#141]
Hi Aurélien, thanks for your awesome book. It helps me a lot and I use it like a reference book. I see you talked about tfa around page 545, and I'm still have some implementation problems about it. I hope you can give me some help.

Can I use a multi layer rnn with the tfa sampler(e.g. two layer lstm with a sampler)? 

Where can I find a full example with schedule sampler?  Can I just replace training sampler in your code with the schedule sampler? I see the schedule sampler use the bernouli probablity, if I use the inverse sigmoid as the sampling probablity, can I give it as the parameter and is it the same length as the timestep length(time step shape).

During inference, should I just use the inference sampler? 

I think the tfa team should have more examples for someone like me who is not familiar with the tf.contrib before.

**Comment by ageron** [comment]
Hi @ruiguo,
Thanks for all your questions. I just pushed the solutions to the coding exercises in chapter 16 (here's [the Colab](https://colab.research.google.com/github/ageron/handson-ml2/blob/master/16_nlp_with_rnns_and_attention.ipynb)), including working examples using TF-Addons' seq2seq API.
Hope this helps (I'll try to answer your questions shortly).

**Comment by ageron** [comment]
> Can I use a multi layer rnn with the tfa sampler(e.g. two layer lstm with a sampler)?

Yes, the encoder must use a `keras.layers.RNN` layer with a `keras.layers.StackedLSTMCell`, itself containing a list of `keras.layers.LSTMCell` instances. The decoder must also use a `keras.layers.StackedLSTMCell`.

> Where can I find a full example with schedule sampler? Can I just replace training sampler in your code with the schedule sampler? I see the schedule sampler use the bernouli probablity, if I use the inverse sigmoid as the sampling probablity, can I give it as the parameter and is it the same length as the timestep length(time step shape).

Yes, just replacing should work. You can start with a low `sampling_probability` (close to 0), train for a few epochs, then increase the sampling probability, train for a few more epochs, and so on. A simple way to do that is to use a `tf.Variable` for the sampling probability, initially set to 0.0, and create a Keras callback that gradually increases the variable at each epoch, eventually close to 1.0.

> During inference, should I just use the inference sampler?

During inference, you should just be able to keep the schedule sampler, but set the sampling_probability to 1.0 so it only samples from the outputs.

> I think the tfa team should have more examples for someone like me who is not familiar with the tf.contrib before.

I agree, there are not enough examples, both for TF 1 and 2.

Hope this helps.

**Comment by ageron** [comment]
Alternatively, it's also possible to code everything yourself using pure tf.keras (with TFA). Check out the [NMT tutorial](https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/text/nmt_with_attention.ipynb), for an example using attention mechanisms and teacher forcing.

**Comment by ageron** [comment]
FYI, I pushed a few examples of how to use TF-Addons' seq2seq API in the notebook 16 (check out the solution to exercise 9). There's an example of using a scheduled sampler, gradually increasing the sampling probability during training, and switching to a greedy embedding sampler afterwards. I also included an example using an attention mechanism and the Keras subclassing API, which makes it nicer to switch from one sampler for training to another for inference.

The example using the scheduled sampler only works in TF 2.2rc3, due to a TF bug that was only resolved recently. Unfortunately, this API broke significantly when migrating to TF 2.0, but they have made a lot of efforts to fix it, and in TF 2.2, it seems to be back to the level I enjoyed in TF 1.x. They still could use a lot more documentation, but hopefully these working examples will help.

**Comment by ruiguo-bio** [comment]
Hi Aurélien, thanks a lot for the solution of question 9! I enjoyed reading your code and help me better understand encoder-decoder model! Now I understood most of that, and will make sure tomorrow if I still have some questions.

**Comment by ruiguo-bio** [comment]
Hi Aurélien. I have some issue about the addon BasicDecoder still, but that is not related to your answer to question 9. I see the code 

`decoder = tfa.seq2seq.basic_decoder.BasicDecoder(decoder_cell,
                                                 sampler,
                                                 output_layer=output_layer)`
if I would like to output multiple output, which is all difference dense layers connected to the shared LSTM sequence output, can I use a list of output_layer in that function? Or should I use multiple BasicDecoder instances, which all connects to a shared LSTM cell instance?
 
Another question is if the true decoder input is not feed into the LSTM cell directly, but first concatenate with the repeated encoder output, which is the same as the 1st solution to the question. Is it possible to do that? When it samples output, how to let the configure the decoder to do other operations first, rather than just use the sampled output to feed into the LSTM cell directly? Thanks!

**Comment by ruiguo-bio** [comment]
It seems I can base that idea on the NMT tutorial you mentioned before, in the train step iterate through the time step. I will try that first!
